### DISCUSSION TITLE
Super Metronome Hero

### DISCUSSION THREAD

**Discussion by djcsdy** [excaliburjs/Excalibur#2522] [Show and tell]
George Weller (@Zutty) and I have been gradually converting some old Flash games to Excalibur.

# [Super Metronome Hero](https://super-metronome-hero.vidja.games/)

![image](https://user-images.githubusercontent.com/310800/203843699-f2cdccfb-e1ee-496b-b735-459c143797b1.png)

George originally built this in Flash using FlashPunk for Fuck This Jam, which was a game jam challenging developers to build games in a genre they dislike. His description:

> Super Metronome Hero is a rhythm game. Rather than trying to make a game I
wanted to play out of this rather popular genre I decided to send it up with
this nonsense joke game. The game has only one button (which is ANY button on
your keyboard, incidentally) which you must press in time at the specified
tempo, every beat for four bars. You unlock different tempos as you progress.
> 
> Rhythm games are the only genre I truly hate, but since I lack any music-making
skills or equipment, I would have had to find decent-quality public domain music
to make a true guitar-hero style game. To save myself this trouble, I decided
instead to boil the concept down to its bare bones. Enjoy!

At the time of its original release this game got a [write-up on CNET](https://www.cnet.com/culture/super-metronome-hero-stops-the-music/).

I did the conversion to Excalibur and also contributed a few small bits to the original game.

* [Super Metronome Hero](https://super-metronome-hero.vidja.games/) - the game itself
* [Source code](https://github.com/dcgw/super-metronome-hero)
* [Article on CNET](https://www.cnet.com/culture/super-metronome-hero-stops-the-music/)

# Other Stuff

Our previous show and tell games are here: https://github.com/excaliburjs/Excalibur/discussions/1845

You can see the other games we're working on at https://github.com/dcgw

**Comment by eonarheim** [comment]
@djcsdy @Zutty) Thanks for sharing! This is so cool to see folks porting things! Let us know if there was something that would have helped the process!

The polish is pretty amazing, really gives me a solid mario party or mario cart vibe! It's also pretty clear I have no rhythm in my body while playing this haha!

**Reply by djcsdy** [reply]
@eonarheim One very minor thing that would have helped: an implementation of `Loadable` for fonts.

I made my own here: https://github.com/dcgw/super-metronome-hero/blob/main/font-loader.ts

See it in use here: https://github.com/dcgw/super-metronome-hero/blob/main/resources.ts#L53-L60 and here: https://github.com/dcgw/super-metronome-hero/blob/main/game.ts#L32

It's not the ideal API, just a quick hack that gets the job done. But if you want, feel free to use/modify/distribute the code however you like, under any license, with or without attribution.

Also mildly annoying is that some implementations of `Loadable` add a cache busting URL parameter by default whereas others don't, and the `bustCache` field became private at some point. This seems like the worst possible implementation, I assume it is an oversight :-). For my case, I use a webpack plugin that adds a hash to the filename, so I don't need a cache busting URL parameter, so I did a little hack to disable it by setting the private `_bustCache` field where it exists here: https://github.com/dcgw/super-metronome-hero/blob/main/resources.ts#L93-L98

**Reply by kamranayub** [reply]
That's true, with bundlers we probably don't need cache-busting 🤔 It was originally added for script-include games...

**Comment by kamranayub** [comment]
Should we be adding these to the Showcase page? 💘

**Reply by djcsdy** [reply]
I'd be honoured if you did :-)

**Reply by eonarheim** [reply]
@djcsdy I just realized we forgot to do this! I'll do this soon!

**Reply by eonarheim** [reply]
@djcsdy Apologies it took me so long but I've added it to the excaliburjs.com/gallery page! Thanks again! 

https://excaliburjs.com/gallery/

**Comment by KokoDoko** [comment]
Thanks for the link to a lot of cool projects, it's very helpful to see how others build their excalibur games.

I noticed the metronome game has a sound loading error? For me the sound doesn't play, and the browser gives this warning:

```
[Warn] :  – "Excalibur was unable to unlock the audio context, audio probably will not play in this browser."
```

I have audio loading problems in my own excalibur game as well!

**Reply by eonarheim** [reply]
I've created a bug for this https://github.com/excaliburjs/Excalibur/issues/2598

**Reply by djcsdy** [reply]
The sound should start playing if you click on the game, although to be honest I've only tested it on Firefox.

It's supposed to have a prompt to tell you to click the game, but I forgot to add it. (I don't like using the play button built into excalibur, which serves the same purpose :-)).

**Reply by djcsdy** [reply]
https://github.com/dcgw/super-metronome-hero/issues/94
Thanks for reporting.
